File: backend/app/api/schemas/actions.py
"""
Action schemas — Pydantic models for the Action system.

Framework alignment (March 2026): Actions are habit or completable items
linked to life domains. Sources: journal_extraction, ai_suggestion, user_created.
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


# ── Enums as literals ─────────────────────────────────────────────

ActionType = Literal["habit", "completable"]
ActionStatus = Literal["active", "paused", "completed", "abandoned"]
ActionSource = Literal["journal_extraction", "ai_suggestion", "user_created"]

VALID_DOMAINS = {"career", "relationship", "family", "health", "finance", "social", "purpose"}


# ── Action schemas ────────────────────────────────────────────────

class ActionCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    action_type: ActionType
    source: ActionSource = "user_created"
    primary_domain: Optional[str] = None
    target_frequency: Optional[int] = Field(None, ge=1, le=7)
    difficulty: Optional[int] = Field(None, ge=1, le=5)

    @field_validator("primary_domain")
    @classmethod
    def validate_domain(cls, v):
        if v is not None and v not in VALID_DOMAINS:
            raise ValueError(f"Invalid domain: {v}. Must be one of {VALID_DOMAINS}")
        return v


class ActionUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    status: Optional[ActionStatus] = None
    primary_domain: Optional[str] = None
    target_frequency: Optional[int] = Field(None, ge=1, le=7)
    difficulty: Optional[int] = Field(None, ge=1, le=5)
    sort_order: Optional[int] = None

    @field_validator("primary_domain")
    @classmethod
    def validate_domain(cls, v):
        if v is not None and v not in VALID_DOMAINS:
            raise ValueError(f"Invalid domain: {v}. Must be one of {VALID_DOMAINS}")
        return v


class ActionResponse(BaseModel):
    id: int
    user_id: int
    title: str
    description: Optional[str] = None
    action_type: str
    status: str
    source: str
    primary_domain: Optional[str] = None
    target_frequency: Optional[int] = None
    difficulty: Optional[int] = None
    sort_order: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ── ActionMilestone schemas ───────────────────────────────────────

class ActionMilestoneCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    sort_order: int = 0


class ActionMilestoneResponse(BaseModel):
    id: int
    action_id: int
    title: str
    is_completed: bool
    completed_at: Optional[datetime] = None
    sort_order: int
    created_at: datetime

    class Config:
        from_attributes = True


# ── HabitLog schemas ─────────────────────────────────────────────

class HabitLogCreate(BaseModel):
    log_date: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$")
    completed: bool = True


class HabitLogResponse(BaseModel):
    id: int
    action_id: int
    user_id: int
    log_date: str
    completed: bool
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/api/actions.ts
/**
 * Actions API client — framework-aligned (March 2026).
 */
import apiClient from './client';
import type { Action, ActionMilestone, HabitLog } from '../types/Action';

// ── Action CRUD ─────────────────────────────────────────────────

export async function createAction(data: {
  title: string;
  description?: string;
  action_type: 'habit' | 'completable';
  source?: string;
  primary_domain?: string;
  target_frequency?: number;
  difficulty?: number;
}): Promise<Action> {
  const res = await apiClient.post('/actions', data);
  return res.data;
}

export async function listActions(params?: {
  status?: string;
  domain?: string;
}): Promise<Action[]> {
  const res = await apiClient.get('/actions', { params });
  return res.data;
}

export async function getAction(actionId: number): Promise<Action> {
  const res = await apiClient.get(`/actions/${actionId}`);
  return res.data;
}

export async function updateAction(
  actionId: number,
  data: Partial<{
    title: string;
    description: string;
    status: string;
    primary_domain: string;
    target_frequency: number;
    difficulty: number;
    sort_order: number;
  }>,
): Promise<Action> {
  const res = await apiClient.patch(`/actions/${actionId}`, data);
  return res.data;
}

export async function deleteAction(actionId: number): Promise<void> {
  await apiClient.delete(`/actions/${actionId}`);
}

// ── Milestones ──────────────────────────────────────────────────

export async function createMilestone(
  actionId: number,
  data: { title: string; sort_order?: number },
): Promise<ActionMilestone> {
  const res = await apiClient.post(`/actions/${actionId}/milestones`, data);
  return res.data;
}

export async function listMilestones(actionId: number): Promise<ActionMilestone[]> {
  const res = await apiClient.get(`/actions/${actionId}/milestones`);
  return res.data;
}

export async function toggleMilestone(
  actionId: number,
  milestoneId: number,
): Promise<ActionMilestone> {
  const res = await apiClient.patch(`/actions/${actionId}/milestones/${milestoneId}`);
  return res.data;
}

export async function deleteMilestone(actionId: number, milestoneId: number): Promise<void> {
  await apiClient.delete(`/actions/${actionId}/milestones/${milestoneId}`);
}

// ── Habit Logs ──────────────────────────────────────────────────

export async function logHabit(
  actionId: number,
  data: { log_date: string; completed?: boolean },
): Promise<HabitLog> {
  const res = await apiClient.post(`/actions/${actionId}/logs`, data);
  return res.data;
}

export async function getHabitLogs(
  actionId: number,
  startDate: string,
  endDate: string,
): Promise<HabitLog[]> {
  const res = await apiClient.get(`/actions/${actionId}/logs`, {
    params: { start_date: startDate, end_date: endDate },
  });
  return res.data;
}
